README.rst

```diff
@@ -92,4 +92,170 @@ By default SQLite memory database will be used.
 .. code-block:: shell
 
     py.test test --test-db-url=sqlite:///test_db.sql
+    py.test test --test-db-url=mysql+mysqlconnector://root:password@localhost:3306/nameko_sqlalchemy_test
+
+Decorators
+----------
+
+transaction_retry
+^^^^^^^^^^^^^^^^^
+This decorator automatically retries the wrapped function when a database connection error occurs.
+If the optional ``session`` argument is passed it will issue a rollback on it before retrying so the transaction can be processed again.
+The ``session`` argument can either be the ``sqlalchemy.orm.session.Session`` or an ``operator.attrgetter`` object if the session is a class attribute.
+
+
+Usage
+"""""
+
+.. code-block:: python
+
+    from sqlalchemy import create_engine
+    from sqlalchemy.orm import sessionmaker
+
+    from nameko_sqlalchemy import transaction_retry
+
+
+    engine = create_engine('postgresql://username:password@localhost/test')
+    Session = sessionmaker(bind=engine)
+    db_session = Session()
+
+    @transaction_retry()
+    def get_example_data():
+        db_session.query(ExampleModel).all()
+
+    example_data = get_example_data()
+
+
+or using with the ``Database`` dependency provider
+
+.. code-block:: python
+
+    from sqlalchemy.ext.declarative import declarative_base
+    from nameko_sqlalchemy import Database, transaction_retry
+
+
+    DeclBase = declarative_base(name='examplebase')
+
+
+    class ExampleService:
+        name = 'exampleservice'
+
+        db = Database(DeclBase)
+
+        @entrypoint
+        @transaction_retry
+        def get_examples(self):
+            with self.db.get_session() as session:
+                return session.query(ExampleModel).all()
+
+        @entrypoint
+        def get_examples_with_retry_inside(self):
+            with self.db.get_session() as session:
+                @transaction_retry(session=session)
+                def foo():
+                    return session.query(ExampleModel).all()
+
+                return foo()
+
+        @entrypoint
+        @transaction_retry
+        def create_example_without_using_context_manager(self):
+            session = self.db.get_session()
+            session.add(ExampleModel(data='hello'))
+            session.commit()
+
+        @entrypoint
+        @transaction_retry(session=operator.attrgetter('db.session'))
+        def create_example_with_worker_scoped_session(self):
+            self.db.session.add(ExampleModel(data='hello'))
+            self.db.session.commit()
+
+.. caution::
+    Using the decorator may cause unanticipated consequences when the decorated function uses more than one session.
```

Comment:
I think it's more correct to say "more than one transaction" here. You'll have the same problem if you do:

``` python
@transaction_retry
def method(self):
      session.add(something)
      session.commit()
      # disconnect/reconnect
      session.add(something_else)
      session.commit()
```

It's probably fair to say quite strongly that this should only be used around single transactions, because all transactions inside the decorator will be re-executed if there is a connection error during any of them.

Reply:
I copy-pasted this to the README, I hope you don't mind 😃